extensions/amp-content-viewer/0.1/amp-content-viewer.js

```diff
@@ -0,0 +1,696 @@
+
+    /** @private {?../../../src/gesture.Gestures} */
+    this.gestures_ = null;
+
+    /** @private {?../../../src/motion.Motion} */
+    this.motion_ = null;
+
+  }
+
+  /** @override */
+  buildCallback() {
+    const children = this.getRealChildren();
+
+    user().assert(children.length == 1, SUPPORT_VALIDATION_MSG);
+    user().assert(
+        this.elementIsSupported_(children[0]),
+        children[0].tagName + ' is not supported by <amp-content-viewer>'
+    );
+    this.content_ = children[0];
+  }
+
+  /** @override */
+  onMeasureChanged() {
+    this.resetContentDimensions_();
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.measureElement(() => {
+      this.sourceWidth_ = this.content_./*OK*/scrollWidth;
+      this.sourceHeight_ = this.content_./*OK*/scrollHeight;
+    });
+    return this.resetContentDimensions_()
+        .then(() => this.setupGestures_());
+  }
+
+  /** @override */
+  pauseCallback() {
+    this.resetContentDimensions_();
+    this.cleanupGestures_();
+  }
+
+  /** @override */
+  resumeCallback() {
+    if (this.content_) {
+      this.scheduleLayout(this.content_);
+    }
+    if (!this.gestures_) {
+      this.setupGestures_();
+    }
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.cleanupGestures_();
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED || Layout.FILL;
+  }
+
+  /** @override */
+  isRelayoutNeeded() {
+    return true;
+  }
+
+  /**
+   * Returns the boundaries of the content element.
+   * @return {?../../../src/layout-rect.LayoutRectDef}
+   */
+  getContentBox() {
+    return this.contentBox_;
+  }
+
+  /**
+   * Returns the content element.
+   * @return {?Element}
+   */
+  getContent() {
+    return this.content_;
+  }
+
+  /**
+   * Returns the boundaries of the content element with the offset if it was
+   * moved by a gesture.
+   * @return {?../../../src/layout-rect.LayoutRectDef}
+   */
+  getContentBoxWithOffset() {
+    if (this.posX_ == 0 && this.posY_ == 0 || !this.contentBox_) {
+      return this.contentBox_;
+    }
+    const expansionScale = (this.scale_ - 1) / 2;
+    return moveLayoutRect(
+        expandLayoutRect(this.contentBox_, expansionScale, expansionScale),
+        this.posX_,
+        this.posY_
+    );
+  }
+
+  /**
+   * Checks to see if an element is supported.
+   * @param {Element} element
+   * @return {boolean}
+   * @private
+   */
+  elementIsSupported_(element) {
+    return ELIGIBLE_TAGS[element.tagName.toLowerCase()];
+  }
+
+  /**
+   * Measures the content viewer and content sizes and positioning.
+   * This must be called AFTER the source element has already been
+   * laid out.
+   * @private
+   */
+  measure_() {
+    this.elementBox_ = layoutRectFromDomRect(this.element
+        ./*OK*/getBoundingClientRect());
+
+    const sourceAspectRatio = this.sourceWidth_ / this.sourceHeight_;
+    let height = Math.min(this.elementBox_.width / sourceAspectRatio,
```

Comment:
This width/height/aspect ratio logic looks familiar from lightbox gallery. Might be worth making a common utility for it if you think it'll be reused again

Reply:
Yup, this is copy-pasta from `amp-image-viewer`. I think this prototype needs to mature a bit more and then we can abstract out the common portions of `amp-X-viewer`.